tests/test_cache_lock_file_perf.py

```diff
@@ -0,0 +1,83 @@
+import logging
+import multiprocessing
+import os
+import sys
+import time
+
+import pytest
+
+from msal_extensions import FilePersistence, CrossPlatLock
+
+
+@pytest.fixture
+def cache_location():
+    path_to_script = os.path.dirname(os.path.abspath(__file__))
+    return os.path.join(path_to_script, str(os.getpid())+"msal.cache")
+
+
+def _validate_result_in_cache(expected_entry_count, cache_location):
+    with open(cache_location) as handle:
+        data = handle.read()
+    prev_process_id = None
+    count = 0
+    for line in data.split(("\n")):
+        if line:
+            count += 1
+            tag, process_id = line.split(" ")
+            if prev_process_id is not None:
+                assert process_id == prev_process_id, "Process overlap found"
+                assert tag == '>', "Process overlap_found"
+                prev_process_id = None
+            else:
+                assert tag == '<', "Opening bracket not found"
+                prev_process_id = process_id
+    return count
+
+
+def _acquire_lock_and_write_to_cache(cache_location, sleep_interval=0):
+    cache_accessor = FilePersistence(cache_location)
+    lock_file_path = cache_accessor.get_location() + ".lockfile"
+    with CrossPlatLock(lock_file_path):
+        data = cache_accessor.load()
+        if data is None:
+            data = ""
+        data += "< " + str(os.getpid()) + "\n"
+        time.sleep(sleep_interval)
+        data += "> " + str(os.getpid()) + "\n"
+        cache_accessor.save(data)
+
+
+def _run_multiple_processes(no_of_processes, cache_location, sleep_interval):
+    open(cache_location, "w+")
+    processes = []
+    for i in range(no_of_processes):
+        t = multiprocessing.Process(
+            target=_acquire_lock_and_write_to_cache,
+            args=(cache_location, sleep_interval))
+        processes.append(t)
+
+    for process in processes:
+        process.start()
+
+    for process in processes:
+        process.join()
+
+
+def test_multiple_processes_without_timeout_exception(cache_location):
+    num_of_processes = 20
+    sleep_interval = 0
+    _run_multiple_processes(num_of_processes, cache_location, sleep_interval)
+    count = _validate_result_in_cache(num_of_processes, cache_location)
+    assert count == num_of_processes * 2
```

Comment:
We tried this earlier in this PR. The proper way to do this is to use:

```suggestion
    if count < num_of_process * 2:
        logging.warning("Starvation detected")
```

Reply:
Yes, but I am seeing in this case it is less than no of processes because of the locking exception and not starvation, if you check [build outputs](https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/58/commits/e19a0b176ef65e54e53b45469e5b1dd45d2bdae8) for this you can verify that,  I was referring to that in our offline teams conversation as well. so if we don't have a timeout, it will work fine.

Reply:
So I introduced that logging of exception in the previous 6 commits to see why the test that is supposed to be running for processes without a timeout is failing because of timeout, if it is then we will have to reduce the num of process even more just to make sure the test actually tests what it is supposed to and does not fail because of timeout

Reply:
You called it "timeout (during lock acquisition)", I called it "starvation". Probably neither is a good name, because naming is hard in computer science. :-)

But the point here is, unable to acquire lock during high load is totally understandable and even expected. That is not an error, and we do NOT reduce the number of processes for that. The goal of this particular test case is NOT to avoid such "abortion due to no lock available" (oops I'm inventing new names again), it is to observe whether garbled data would occur when lots of processes accessing the same cache.

Whether/how to detect "abortion due to no lock available" is a different topic. You are welcome to still attempt it, based on all the understandings above.

Reply:
I think it does make sense to rename the test names to distinguish the purpose because in both the cases we are ok having a timeout, its just testing them on high load or low load and finding the behaviour only. This tests will have no relation to testing of locks then as that behaviour does not even kick in case of or library because of timeouts.

Reply:
Having said that we can always log the timeout exception to see what hapenned there. So I might anyway add that and then may be we add one more test case which succeeds without any difficulty under normal load

Reply:
> we can always log the timeout exception to see what hapenned there. So I might anyway add that

No problem. Just make sure you specifically catch the exception raised by portalocker. Do not just `except Exception`.

> may be we add one more test case which succeeds without any difficulty under normal load
That is also welcome. Make sure we will have 3 meaningful names. :-)

Reply:
> This tests will have no relation to testing of locks then as that behaviour does not even kick in case of or library because of timeouts.

Not quite. When that "timeout (during lock acquisition)" happens, it is exactly the locking mechanism doing its job. And, even when in usual cases when everything seems just working fine, that is still the locking mechanism doing its job on guarding race conditions.

Reply:
Sure, that is correct, what I meant is that the performance tests that Petro found issues with was with a really long timeout and then checking how to the lock behaves and thats where we found a few issues on Mac. These tests are definitely not testing that behavior.